File: kurier.py
import sys
import heapq

def main():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
    
    n = int(input_data[0])
    m = int(input_data[1])
    s = int(input_data[2])
    t = int(input_data[3])
    
    adj = [[] for _ in range(n + 1)]
    idx = 4
    for _ in range(m):
        u = int(input_data[idx])
        v = int(input_data[idx+1])
        w = int(input_data[idx+2])
        adj[u].append((v, w))
        adj[v].append((u, w)) 
        idx += 3
        
    dist = [float('inf')] * (n + 1)
    min_nodes = [float('inf')] * (n + 1)
    
    dist[s] = 0
    min_nodes[s] = 1 
    
    pq = [(0, 1, s)]
    
    while pq:
        d, nodes, u = heapq.heappop(pq)
        
        if d > dist[u] or (d == dist[u] and nodes > min_nodes[u]):
            continue
            
        for v, weight in adj[u]:
            if dist[u] + weight < dist[v]:
                dist[v] = dist[u] + weight
                min_nodes[v] = nodes + 1
                heapq.heappush(pq, (dist[v], min_nodes[v], v))
            elif dist[u] + weight == dist[v]:
                if nodes + 1 < min_nodes[v]:
                    min_nodes[v] = nodes + 1
                    heapq.heappush(pq, (dist[v], min_nodes[v], v))
                    
    if dist[t] == float('inf'):
        print("NIE")
    else:
        print(dist[t], min_nodes[t])

if __name__ == '__main__':
    main()

File: kurier.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

const long long INF = 1e18;

struct State {
    long long time;
    int nodes;
    int u;

    bool operator>(const State& other) const {
        if (time != other.time) return time > other.time;
        return nodes > other.nodes;
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m, s, t;
    if (!(cin >> n >> m >> s >> t)) return 0;
  
    vector<vector<pair<int, long long>>> adj(n + 1);
    for (int i = 0; i < m; ++i) {
        int u, v;
        long long w;
        cin >> u >> v >> w;
        adj[u].push_back({v, w});
        adj[v].push_back({u, w}); 
    }

    vector<long long> dist(n + 1, INF);
    vector<int> min_nodes(n + 1, 1e9);

    dist[s] = 0;
    min_nodes[s] = 1; 

    priority_queue<State, vector<State>, greater<State>> pq;
    pq.push({0, 1, s}); 

    while (!pq.empty()) {
        auto [d, nodes, u] = pq.top();
        pq.pop();

        if (d > dist[u] || (d == dist[u] && nodes > min_nodes[u])) continue;

        for (auto edge : adj[u]) {
            int v = edge.first;
            long long weight = edge.second;

            if (dist[u] + weight < dist[v]) {
                dist[v] = dist[u] + weight;
                min_nodes[v] = nodes + 1;
                pq.push({dist[v], min_nodes[v], v});
            } 
            else if (dist[u] + weight == dist[v]) {
                if (nodes + 1 < min_nodes[v]) {
                    min_nodes[v] = nodes + 1;
                    pq.push({dist[v], min_nodes[v], v});
                }
            }
        }
    }

    if (dist[t] == INF) {
        cout << "NIE\n";
    } else {
        cout << dist[t] << " " << min_nodes[t] << "\n";
    }

    return 0;
}
